technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,122 @@
+Currently Medusa is not dc/rack aware and it can probably be fairly surprising to users. It was built to backup a single DC per cluster and didn’t account for racks either.  
+Restoring from a DC/rack aware cluster to another one imposes some constraints on the source/target nodes mapping, which are currently impossible to deal with safely in a remote restore: 
+
+When using NetworkTopologyStrategy, the placement of replicas is dependent on the DC and rack layout.
+In this context, the replica for a token range will be placed on the node that owns the next token range in the next rack.
+We can't restore any node to any other node as we may end up putting several replicas in the same rack, and the nodes we'd restore to would possibly end up not containing data they're supposed be replica for.
+We can be rigid about the mapping and expect the same dc and rack names exclusively, or do a 1:1 mapping between DCs/Racks despite having different names (but expect the same number of both).
+
+
+## Restore Orchestration
+
+1. Get backup definition from storage backend
+![](img/K8ssandra-new-restore-design-1.jpg)  
+The restore operation will assume the CRDs have been synced with the remote storage backend. Backup topology (hostnames and tokens) will be retrieved from the new CRD format.
+
+
+2. Set the env variables on the restore containers
+![](img/K8ssandra-new-restore-design-2.jpg)  
+The backup definition will be used to map the backup hosts with the restore hosts and perform checks on whether or not the restore is possible:
+
+  - Same number of DCs
+  - Same number of nodes per DC
+  - Same rack distribution per DC
+  
+  Nodes will be mapped 1:1 between the source and destination topologies (sorted alphabetically). If the mapping fails to associate nodes 1:1, the restore operation must fail.
+
+  The init containers will get the following env variables set:
+
+  - Backup name
+  - Restore key
+  - Bucket prefix (new)
+  - In Place/Remote (new)
+
+  These values will then be used to invoke medusa.restore_node in the medusa-restore init container.
+The init-container needs to deal with mapping between source and dest topologies to self determine which backup it needs to restore. Pre-checks in the operator will guarantee that a 1:1 mapping is possible and topologies are compatible.
+
+3. Shutdown Cassandra to trigger the restore
+![](img/K8ssandra-new-restore-design-3.jpg)  
+
+  medusa-operator will patch the cassandradatacenter spec to provoke a shutdown of the Cassandra nodes through cass-operator. The pods will be restarted and the medusa-restore init container will trigger the restore.
+
+4. Restore the backups
+
+![](img/K8ssandra-new-restore-design-4.jpg)  
+
+  The medusa-restore container will perform a remote restore which involves the following:
+system.local and system.peers will not be restored
+auto_bootstrap will be set to false and seed nodes will start first
+tokens will be enforced by modifying cassandra.yaml
```

Comment:
Interesting, so we are going to override the current token distribution and apply the distribution from the origin cluster. Indeed, this would require all nodes to be stopped at once. Couldn't we image a softer approach: same token distribution in origin and target => no need to do a full shutdown. For people with strong SLAs, avoiding a downtime during restore could be a good thing, unless this procedure is being designed primarily for setting up new clusters from scratch (as opposed to restoring a backup to an existing, running cluster).

Reply:
We're working on live restores for in place restores only (although we'd still prevent apps from accessing the cluster while the restore is running). Here we're talking about remote restores which will most likely require to enforce the tokens.
Also Medusa has an "all or nothing" philosophy. While people could expect to be able to restore a single table from a cluster and a single keyspace from another one to the same target cluster, that comes with too many challenges to be reliable in the Cassandra world. Stop, wipe and restore is how we do stuff in Medusa, for better or worse :)